src/swap/handle_swap_sign_transaction.c

```diff
@@ -136,79 +206,73 @@ bool swap_check_validity() {
         os_sched_exit(0);
     }
 
-    if (G_context.tx_info.transaction.hints_type != TRANSACTION_COMMENT) {
+    if ((G_context.tx_info.transaction.hints_type == TRANSACTION_COMMENT) &&
+        (G_context.tx_info.transaction.hints_len != 0)) {
+        PRINTF("Hint length %d refused\n", G_context.tx_info.transaction.hints_len);
+        io_send_sw(SW_SWAP_FAILURE);
+        // unreachable
+        os_sched_exit(0);
+    } else if ((G_context.tx_info.transaction.hints_type != TRANSACTION_COMMENT) &&
+               (G_context.tx_info.transaction.hints_type != TRANSACTION_TRANSFER_JETTON)) {
         PRINTF("Wrong operation %d\n", G_context.tx_info.transaction.hints_type);
         io_send_sw(SW_SWAP_FAILURE);
         // unreachable
         os_sched_exit(0);
-    } else if (G_context.tx_info.transaction.hints_len != 0) {
-        PRINTF("Hint length %d refused\n", G_context.tx_info.transaction.hints_len);
+    }
+
+    if ((strncmp(G_swap_validated.ticker, "TON", sizeof("TON")) == 0) &&
+        (G_context.tx_info.transaction.hints_type == TRANSACTION_TRANSFER_JETTON)) {
+        PRINTF("Wrong operation type %d for native coin swap\n",
+               G_context.tx_info.transaction.hints_type);
         io_send_sw(SW_SWAP_FAILURE);
         // unreachable
         os_sched_exit(0);
-    } else {
-        PRINTF("Valid operation %d\n", G_context.tx_info.transaction.hints_type);
     }
 
+    PRINTF("Valid operation %d\n", G_context.tx_info.transaction.hints_type);
+
     if (G_context.tx_info.transaction.send_mode & 128) {
         PRINTF("Amount MAX is refused\n");
         io_send_sw(SW_SWAP_FAILURE);
         // unreachable
         os_sched_exit(0);
     }
 
-    if (G_swap_validated.amount_length != G_context.tx_info.transaction.value_len) {
+    uint8_t amount_length = G_context.tx_info.transaction.value_len;
+    uint8_t* amount = G_context.tx_info.transaction.value_buf;
+    // Depending on the hints type, we need to get the proper amount from the transaction
+    if (G_context.tx_info.transaction.hints_type == TRANSACTION_TRANSFER_JETTON) {
```

Comment:
Same as before, I think we should set a `bool is_jetton_swap` at function entrance by using the ticker check, and then reuse it as source of truth